actions/bcr-pr-review-approver/index.js

```diff
@@ -0,0 +1,262 @@
+
+  for (const module of modifiedModules) {
+    let moduleApproved = false;
+    for (const [maintainer, maintainedModules] of maintainersMap.entries()) {
+      if (maintainedModules.has(module) && approvers.has(maintainer)) {
+        moduleApproved = true;
+        console.log(`Module '${module}' has maintainers' approval from '${maintainer}'.`);
+        break;
+      }
+    }
+    if (!moduleApproved) {
+      allModulesApproved = false;
+      modulesNotApproved.push(module);
+      console.log(`Module '${module}' does not have maintainers' approval.`);
+    }
+  }
+
+
+  if (!allModulesApproved) {
+    console.log(`Cannot auto-merge this PR, the following modules do not have maintainers' approval: ${modulesNotApproved.join(', ')}`);
+  } else {
+    console.log('All modified modules have maintainers\' approval');
+  }
+
+  return allModulesApproved;
+}
+
+async function reviewPR(octokit, owner, repo, prNumber) {
+
+  console.log('\n');
+  console.log(`Processing PR #${prNumber}`);
+
+  // Fetch modified modules
+  const modifiedModules = await fetchAllModifiedModules(octokit, owner, repo, prNumber);
+  console.log(`Modified modules: ${Array.from(modifiedModules).join(', ')}`);
+  if (modifiedModules.size === 0) {
+    console.log('No modules are modified in this PR');
+    return;
+  }
+
+  // Figure out maintainers for each modified module
+  const [ maintainersMap, modulesWithoutGithubMaintainers ] = await generateMaintainersMap(octokit, owner, repo, modifiedModules);
+  console.log('Maintainers Map:');
+  for (const [maintainer, maintainedModules] of maintainersMap.entries()) {
+    console.log(`- Maintainer: ${maintainer}, Modules: ${Array.from(maintainedModules).join(', ')}`);
+  }
+
+  // If modulesWithoutGithubMaintainers is not empty, then return
+  if (modulesWithoutGithubMaintainers.size > 0) {
+    console.log(`Cannot auto-merge this PR with maintainers approval because the following modules do not have maintainers with GitHub usernames: ${Array.from(modulesWithoutGithubMaintainers).join(', ')}`);
+    return;
+  }
+
+  // Get the approvers for the PR
+  const approvers = await getPrApprovers(octokit, owner, repo, prNumber);
+
+  // Verify if all modified modules have at least one maintainer's approval
+  const allModulesApproved = await checkIfAllModifiedModulesApproved(modifiedModules, maintainersMap, approvers);
+
+  const myLogin = context.actor;
+
+  // Approve the PR if not previously approved and all modules are approved
+  if (allModulesApproved) {
+    if (!approvers.has(myLogin)) {
+      console.log('Approving the PR');
+      await octokit.rest.pulls.createReview({
+        owner,
+        repo,
+        pull_number: prNumber,
+        event: 'APPROVE',
+        body: 'Helllo @bazelbuild/bcr-maintainers, all modules in this PR have been approved by their maintainers. Please take a final look to merge this PR.',
+      });
+    }
+
+    // TODO: Enable the following when it's safe to do so.
```

Comment:
is there an API to simply turn on autosubmit? maybe the checks haven't finished yet, or are blocked because it changes presubmit.yml

Reply:
Unfortunately not with the GitHub rest APIs. We can run this GitHub action every half an hour, so it should not be delayed for too long.

https://github.com/orgs/community/discussions/24719